Allow extends-clauses in package.order. (#3635)
modelica/ModelicaSpecification: chapters/packages.tex

A directory shall not contain both the sub-directory \filename{A} and the file \filename{A.mo}.
\end{example}

In order to preserve the order of sub-entities it is advisable to create a file \filename{package.order} where each line contains the name of one class, constant (using its Modelica \lstinline!IDENT! form), or \lstinline!extends!-clause (in the form \lstinline!extends A.B! -- one space, no semicolon).
If a \filename{package.order} is present when reading a structured entity the classes, constants, and optionally \lstinline!extends!-clauses are added in this order; if the contents does not exactly match the classes and constants in the package, the resulting order is tool specific and a warning may be given.

> The relative order of classes, constants, and optionally \lstinline!extends!-clauses that are stored in \filename{package.mo} must be consistent with the relative order of the corresponding entries in \filename{package.order} (that is, the \filename{package.order} only adds information regarding the ordering of classes not stored in \filename{package.mo}).

Reviewer: Isn't this sentence meant as a replacement for the sentence below?

Classes and constants that are stored in \filename{package.mo} are also present in \filename{package.order} but their relative order should be identical to the one in \filename{package.mo} (this ensures that the relative order between classes and constants stored in different ways is preserved).

> An \lstinline!extends!-clause entry shall be omitted from \filename{package.order} if it would appear before any class not stored in \filename{package.mo}.

Reviewer: This rule is so complicated that I'm afraid few humans will ever learn it, and by not even allowing the obvious natural thing of adding every `extends`-clause, there is no migration path to a simple logic that even a human could remember.  Shouldn't we strive for something simpler that also humans can learn as the main rule, and use deprecation to still permit the complicated rule?

Author: My order of preference is thus:
1. shall or should, I can live with either.
2. reject proposal
3. "may as deprecated behavior"
4. may

Author: Off-line poll.

Reviewer: > The downside by deprecating the current behavior is that we force all libraries to be upgraded, and that will be too much pain for very little benefit, as packages regularly have extends-clauses, but rarely want them later. I would say that not changing package.order is preferable to that.

The point of deprecation is that it does _not_ force all libraries to be upgraded anytime soon.  We could have it deprecated for ten years, effectively just recommending to do it the other way for the foreseeable future.  Then, when we have given libraries ample time to adapt, we can remove the deprecated behavior completely.

> If an \lstinline!extends!-clause is missing from \filename{package.order} it shall be added at the first position consistent the order in \filename{package.mo}.

Author: ```suggestion
If an \lstinline!extends!-clause is missing from \filename{package.order} it shall be added at the first position consistent with the order in \filename{package.mo}.
```

In order to preserve the order of sub-entities it is advisable to create a file \filename{package.order} where each line contains the name of one class or constant (using its Modelica \lstinline!IDENT! form).
If a \filename{package.order} is present when reading a structured entity the classes and constants are added in this order; if the contents does not exactly match the classes and constants in the package, the resulting order is tool specific and a warning may be given.
Classes and constants that are stored in \filename{package.mo} are also present in \filename{package.order} but their relative order should be identical to the one in \filename{package.mo} (this ensures that the relative order between classes and constants stored in different ways is preserved).

> Any extends-clause (stored in \filename{package.mo}) may be present in \filename{package.order} in the form \lstinline!extends A.B! (one space, no semicolon), but should be omitted if the extends-clause(s) precede normal classes.

Reviewer: The main form is the singular:
```suggestion
Any extends-clause (stored in \filename{package.mo}) may be present in \filename{package.order} in the form \lstinline!extends A.B! (one space, no semicolon), but should be omitted if the extends-clause(s) precedes normal classes.
```

Reviewer: More complete suggestion addressing additional concerns in review comment:
```suggestion
A \lstinline!extends!-clause present in \filename{package.mo} may be listed in \filename{package.order} in the form \lstinline!extends A.B! (one space, no semicolon), except at the start of \filename{package.order}.
That is, the \filename{package.order} must contain at least one entry without \lstinline!extends! before the first \lstinline!extends!-clause entry.
\begin{nonnormative}
The rule regarding entries at the start of \filename{package.order} offers full backwards compatibility with older packages in the common case of all \lstinline!extends!-clauses appearing first in \filename{package.mo}.
\end{nonnormative}
```

Author: The suggestion of not allowing at the start doesn't fully work, since there can also be constants in package.order. 

Since they cannot be stored in separate files they must be in package.mo - and even if usually the extends-clause precede all of them it is possible to have a mix, but as long as the extends-clause precedes any _class_ there is no need to include the extends in package.order.

Reviewer: I think that this is better addressed by my proposal below in https://github.com/modelica/ModelicaSpecification/pull/3635#issuecomment-2575008137.  According to that proposal, the problematic packages would just end falling into the deprecated behavior.
